src/mbgl/style/expression/value.cpp

```diff
@@ -0,0 +1,322 @@
+#include <rapidjson/writer.h>
+#include <rapidjson/stringbuffer.h>
+#include <mbgl/style/expression/value.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+type::Type typeOf(const Value& value) {
+    return value.match(
+        [&](bool) -> type::Type { return type::Boolean; },
+        [&](double) -> type::Type { return type::Number; },
+        [&](const std::string&) -> type::Type { return type::String; },
+        [&](const mbgl::Color&) -> type::Type { return type::Color; },
+        [&](const NullValue&) -> type::Type { return type::Null; },
+        [&](const std::unordered_map<std::string, Value>&) -> type::Type { return type::Object; },
+        [&](const std::vector<Value>& arr) -> type::Type {
+            optional<type::Type> itemType;
+            for (const auto& item : arr) {
+                const type::Type t = typeOf(item);
+                const std::string tname = type::toString(t);
+                if (!itemType) {
+                    itemType = {t};
+                } else if (type::toString(*itemType) == tname) {
+                    continue;
+                } else {
+                    itemType = {type::Value};
+                    break;
+                }
+            }
+            
+            if (!itemType) { itemType = {type::Value}; }
+
+            return type::Array(*itemType, arr.size());
+        }
+    );
+}
+
+void writeJSON(rapidjson::Writer<rapidjson::StringBuffer>& writer, const Value& value) {
+    value.match(
+        [&] (const NullValue&) { writer.Null(); },
+        [&] (bool b) { writer.Bool(b); },
+        [&] (double f) {
+            // make sure integer values are stringified without trailing ".0".
+            f == std::floor(f) ? writer.Int(f) : writer.Double(f);
+        },
+        [&] (const std::string& s) { writer.String(s); },
+        [&] (const mbgl::Color& c) { writer.String(c.stringify()); },
```

Comment:
- For `Color`, I think just because I'd seen it qualified in a number of other places in the codebase, so was following suit.
- `mbgl::Value` is qualified to distinguish it from `mbgl::expression::Value`
- For the expression types, it seemed to me like having the `type::` qualifier added to clarity.  I can remove it if others disagree.